content/en/blog/_posts/2020-08-04-leaving-the-swarm-road-to-kubernetes.md

```diff
@@ -0,0 +1,172 @@
+
+#### What is cmdr?
+
+In order to manage the bootstrap process of standing up the local environment, I created a Python application, `cmdr`, that handles configuring kind, installing [Traefik](https://traefik.io), [MetalLB](https://metallb.universe.tf/), and deploying our infrastructure, application, and UI components through [Helm](https://helm.sh). We had already standardized how we name, build, and configure our services so everything was already abstracted and DRY - which makes creating new services fairly painless for the developers on the Single Music team. 
+
+In addition to the local bootstrap, I created a `deploy` command accepting a project name, which is referenced inside an over-arching `projects.yaml` file, which then extracts environment-specific details (ingress endpoints, environment configuration) to apply to the correlating helm chart for that project. We use several open-source helm charts for infrastructure components such as MySQL, Postgres, Redis, RabbitMQ, and Confluent Cloud (Kafka). The deploy command also accepts env flags, image tags, application versions, all of which configure metadata and chart specific version tags which help us track application versions across our environments in EKS.
+
+The developers also built a command `promote` that promotes applications from our staging environment through a GitLab pipeline (our CI is the only agent capable of deploying to our cluster). At this point, we were ready to stand up our service in the cloud running on a managed Kubernetes environment. We chose Amazon EKS as we were already running in that provider and had just purchased several thousand dollars worth of reserved instances. We also store all of our customers’ music securely in S3 and depend on their other data services, such as Redshift, and RDS, already. Vendor lock-in is real, y’all.
+
+The `cmdr` tool has been [released](https://github.com/notsureifkevin/cmdr) for anyone to use on their own projects, as I’ll be using it for workshops and demos in the future.
+
+### Step 2: The Migration
+
+Single Music processes roughly 1.5 million transactions every day. Having an extended or failed migration would be a disaster. In addition to building the environment developers could work in, we had to ensure that we were using compatible infrastructure components as well -- which included upgrading from [Traefik 1.7 to 2.2](https://docs.traefik.io/migration/v1-to-v2/), leveraging their [new CRDs](https://docs.traefik.io/v2.0/reference/dynamic-configuration/kubernetes-crd/), and utilizing the Bitnami [Redis](https://github.com/bitnami/charts/tree/master/bitnami/redis) and [RabbitMQ](https://github.com/bitnami/charts/tree/master/bitnami/rabbitmq) charts so our caching and messaging systems are in a position to be scalable and highly available in the not-so-distant future.
```

Comment:
```suggestion
Single Music processes roughly 1.5 million transactions every day. Having an extended or failed migration would be a disaster. In addition to building the environment developers could work in, we had to ensure that we were using compatible infrastructure components as well — which included upgrading from [Traefik 1.7 to 2.2](https://docs.traefik.io/migration/v1-to-v2/), leveraging their [new CRDs](https://docs.traefik.io/v2.0/reference/dynamic-configuration/kubernetes-crd/), and utilizing the Bitnami [Redis](https://github.com/bitnami/charts/tree/master/bitnami/redis) and [RabbitMQ](https://github.com/bitnami/charts/tree/master/bitnami/rabbitmq) charts so our caching and messaging systems are in a position to be scalable and highly available in the not-so-distant future.
```
(just fixing the em dash)
or
```suggestion
Single Music processes roughly 1.5 million transactions every day. Having an extended or failed migration would be a disaster. In addition to building the environment developers could work in, we had to ensure that we were using compatible infrastructure components as well — which included upgrading from [Traefik 1.7 to 2.2](https://docs.traefik.io/migration/v1-to-v2/), leveraging their [new CustomResourceDefinitions](https://docs.traefik.io/v2.0/reference/dynamic-configuration/kubernetes-crd/), and utilizing the Bitnami [Redis](https://github.com/bitnami/charts/tree/master/bitnami/redis) and [RabbitMQ](https://github.com/bitnami/charts/tree/master/bitnami/rabbitmq) charts so our caching and messaging systems are in a position to be scalable and highly available in the not-so-distant future.
```

(in case the reader doesn't know CRD as an abbreviation)